### PR TITLE
Upgraded to 0.54.1 Flow

### PR DESCRIPTION
**Pull request by gforge**
There were some changes with the 0.53 version of flow. The props should now go in the `React$Component<Props>` and there is a [new class](https://flow.org/en/docs/react/types/#toc-react-componenttype) called `React$ComponentType`

### Checklist:

- [x] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

**Comment by gforge** [comment]
I think the PR is fine but it's easy to break the tests, e.g. adding loading gives an error:

```javascript
const withData: OperationComponent<IQuery> = graphql(query);

// Test a functional component
const FunctionalWithData = withData(({ data: { foo, loading } }) => {
  // $ExpectError string type beeing treated as numerical
  if (foo > 1) return <span />;

  return null;
});
```

I also think that there are strange things happening when using a React component with intersections

**Comment by jbaxleyiii** [comment]
@gforge with these changes in floe 0.53, I think we need to tweak the typings to not require mutate and data for each props set. Can you take a look at that?

**Comment by gforge** [comment]
@jbaxleyiii unfortunately this seems to be a little tricky... The natural way would be to return a union and changing the `OperationComponent` to `QueryComponent` and `MutateComponent` like this:

```js
export type ChildQueryProps<P, R> = {
  data: QueryProps & R,
} & P;

export interface QueryComponent<
  TResult: Object = {},
  TOwnProps: Object = {},
  TMergedProps = ChildQueryProps<TOwnProps, TResult>,
> {
  (
    component: React$ComponentType<TMergedProps>,
  ): React$ComponentType<TOwnProps>,
}

export type ChildMutateProps<P, R> = {
  mutate: MutationFunc<R>,
} & P;

export interface MutateComponent<
  TResult: Object = {},
  TOwnProps: Object = {},
  TMergedProps = ChildMutateProps<TOwnProps, TResult>,
> {
  (
    component: React$ComponentType<TMergedProps>,
  ): React$ComponentType<TOwnProps>,
}

type ReturnComponent<TResult, TProps, TChildProps> = 
  QueryComponent<TResult, TProps, TChildProps> | MutateComponent<TResult, TProps, TChildProps>;
declare export function graphql<TResult, TProps, TChildProps>(
  document: DocumentNode,
  operationOptions?: OperationOption<TProps, TResult>,
): ReturnComponent<TResult, TProps, TChildProps>;
```

Unfortunately this triggers for me an internal error `. Internal error: Failure(add_output: no source for error: EIncompatibleDefs { reason_lower = : "Object"; reason_upper = -:99:22-58: "extends QueryComponent"; extras = _ })`. this can be remedied by changing the return type of `graphql` to `any` or just one of the union components. Both of these seem worse options than just adding a mutate option to the child component. I have set-up [an example](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAIKQA2XEAolGJQ4AXhQA7JoxZsAYlhi4AFgFloWAIoBXLFCYB5MDGARx6McknTWHbr226m2VFq7wLV5jfmLV6w8am5hJSXmwAyloARqi4UMBRWAAqEGqUASZmoiHWbACqYAAmyDCaOnoZQdmWoTJwACIQuFogWOIwAHIQhWwetTaGuiWZleLVVnRYAB6QsHBhcCpaMMOmslriuAA8Sc6uMAB82QAUdHBwEEaoAFwhADR0AJSiRzX0kzPQ8MDtugR4bAcegACsQwOgAN5nOC6IQAflunAgPH4gmgD3O4gUAHdoABrcIrGBaG5wcQtRJQDFwXjIQo-ADmtyiEGROHE1IAbsh4sgolwsKT9FEAFZYfDUggKZRpLC3U7nc5S3yy0aknwy9RAgxGTLoABkcA1fnSuqC1OeIiOoJIwFQWC2SJR2r2bi2NQOB2plGVynl3N5-MFCLgwrF+Et1uIIDtDqd9nKTkF+3dkk91NQK1gwORXEZ8rsAElflBuVxbuSQJTI3BORBgIUMzBLjmePm4Mca3WGxnorF4okUrL5ZdAmZbpEYnEEslUv4zWYa52XrX643oVoiiUyo55dDziBkGBZOIC5ROSYSS7k25bjU7hcF6SCsVStq1TX79Cu2uHgBfaEPlmeAFgAYSUYAuEKbUbXBLZgQfDAjjEKFzlfZBbhgsEDUwf84ENYF3mmYC4B+UooH+XBAUTUCSEgLF2i2aFdhvGBbjDcV3DgCE-2pJJ9GxcRYKFUVOOyHi+JUXQGSwQphOycDIOgxNhJ2AShOwh8WJcNwvToI5UI7fc4FwOjTDaNjMCoGAABJaPAcz2iSGQdikqAZLk7C9POR5bmwPBbPs+iLOc1g1ME4S9IAoCvnmOpFKgpYiSwVSEMwZDuOhEBlm3W4ktWcR1k2LYkLwgj3hiuYyL+AFFhy0ogscmAmPObT9nY0T8HE3jmPU4SOvDLiJOYtyPPksQEsKfLSlU-iIs0uA2t0h4DOhBVFVMhyGMs-z8DsszttCh0klG2TIotPzrP2raQpcuaNMuVAooYBZsGJKBxEa7adldGAtOErTJsi7JoW1L6LJ+1j-oWpIga8uAAB86uS8HGKWv7FoBxa4ceg56B6XAuB5NhiNiggNnwTI4AZKAjyUABHLhIZ0jGkix2GIKgyL1sKJoWgs25GmaVp2i6HpqUuIYxzVENBlp6WFx2dnfr03yQneBZC21cToQIVlbkzeJxAZakoh5A2YCNk26D-eg4DoAB6B2OBMonUHQTbgvaOBsWAGAlFImAyHQCBBNrHlgD5AVcjgUDwC4KY+uw8S4H+AAvC2rbgW2Y7jsAE9RmBxsytCSgwuAtcTfC4Cwx7qQAOkbmy+CmAKtjzhOk9x-96EJtB0A7qZC5hKZSnEQp0F2mB68L9v46Hg6LOBwzKHH3Rl0M85TLMeAITgdDbj32l6WNh9YXROAzapVPWWzh90+z7J-bteuwGwu3FWAAgO2Pxlnkod6Ywtj0k5AcAAMhAOkjItgOxAXjYyX8OznygP-BQWgPpwC2EoAAjAcPgGAMD6AwDAnB8DjJOzgM3GYnEBBCDgIbRkcUbCJCwIwy2OBSiFBQOgSsuhgC4GQFwBB39jhXzgEcbBqDAFklcFwd4io4AUKSEoNgiReDYh9vidAaAR6sHwLJYyAD0FjHWgo4BwBQHGQURCPWEB64CmNv7P89DLaMIcQyf2Jk6LEy4c2VOyA07cXTvYtoHilDOPcf7KxmDYEWIOMZR4H8AIAXOHQbemY4D00THeSQ2Rg4kDYFkxwtA0lBHgL7f2shLC4CYPUMumEaKL0YpXRwD5B5d3BBlGmdNGbHCKXoRJpSd6x3ngAdT9koOpKxsgVKUFUzYtSy7HEHoXQZQA) in the Flow *try*. There also seems to be some additional issues with that example.

I've actually always found it a little weird that `graphql` does both mutate and query. A more intuitive approach would be to have `addQuery`, `addMutation` or some verb that separates them - this is though way beyond the scope of the current task.

**Comment by arahansen** [comment]
Are there any additional tasks left for this PR? Or anything I can help take a look at? Would love to see this merged 😁

**Comment by jbaxleyiii** [comment]
@arahansen I partially wanted to wait for the flow react changes to stabilize (for instance .54 is being tested now) but I can work to get it out this week

File: src/index.js.flow [hunk 1]
```diff
@@ -131,10 +128,8 @@ declare export function graphql<TResult, TProps, TChildProps>(
 ): OperationComponent<TResult, TProps, TChildProps>;
 
 declare export function withApollo<TProps>(
-  component:
-    | StatelessComponent<TProps & ApolloClient>
-    | Class<React$Component<any, TProps & ApolloClient, any>>,
-): Class<React$Component<void, TProps & ApolloClient, void>>;
+  component: React$ComponentType<TProps & ApolloClient>,
```
Comment (calebmer) [inline src/index.js.flow]: You might need to make the intersection the other order. i.e. `ApolloClient & TProps`. Given how we implement intersections in Flow, order matters 😊

File: test/flow.js [hunk 2]
```diff
@@ -34,19 +27,99 @@ const mutation = gql`
...
+const FunctionalWithData = withData(({ data: { foo } }) => {
+  // $ExpectError string type beeing treated as numerical
   if (foo > 1) return <span />;
 
   return null;
 });
 
+// Test class component, this requieres a stricter definition
+type BasicComponentProps = {
+  data: IQuery & QueryProps,
```
Comment (calebmer) [inline test/flow.js]: You could provide some utility type like `GraphQLProps<IQuery>` and recommend users spread that in their props type:

```
type Props = {
  ... GraphQLProps<IQuery>,
  myProp: number,
};
```

This also means you could include `mutate` as mixed. For mutations you could have a different type like `GraphQLMutationProps<IMutation>`.

File: src/index.js.flow (hunk 1 above)
Comment (gforge) [reply src/index.js.flow]: Fixed - the test for withApollo is though rather crude, not sure it's correctly implemented

File: test/flow.js (hunk 2 above)
Comment (gforge) [reply test/flow.js]: Tried, can't get it to work :-(

Comment (jbaxleyiii) [reply test/flow.js]: @calebmer would you be able to help out here? I won't be able to try the upgrade today

**Comment by Siyfion** [comment]
I was just coming to the Issues section to report this issue I was experiencing; glad to see that people are already well on-it and progress is underway.

**Comment by gforge** [comment]
Merged with upstream and added the 0.54.1 - can we merge this? It seems like the remaining issues could move to a separate issue

**Comment by chollier** [comment]
that would be great

**Comment by FezVrasta** [comment]
Will this PR fix this error as well?

```
Flow validation
Library type error:
node_modules/react-apollo/graphql.js:0
inconsistent use of library definitions
 86:                 var client = mapPropsToOptions(nextProps).client;
                                                    ^^^^^^^^^ unused function argument
   48:         mapPropsToOptions = function () { return options; };
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ function expects no arguments
```

**Comment by jbaxleyiii** [comment]
@gforge merging and releasing this today!

**Comment by gforge** [comment]
@jbaxleyiii thanks!

@FezVrasta not sure, have you checked against this branch?

**Pull request merged**